internal/provider/resource_gitlab_project_protected_environment.go

```diff
@@ -0,0 +1,198 @@
+package provider
+
+import (
+	"context"
+	"log"
+	"net/http"
+
+	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/validation"
+	gitlab "github.com/xanzy/go-gitlab"
+)
+
+// https://docs.gitlab.com/ee/ci/environments/protected_environments.html
+var _ = registerResource("gitlab_project_protected_environment", func() *schema.Resource {
+	// lintignore: XR002 // TODO: Resolve this tfproviderlint issue
+	return &schema.Resource{
+		Description: "This resource allows you to create and manage Project Access Token for your GitLab projects.",
+
+		CreateContext: resourceGitlabProjectProtectedEnvironmentCreate,
+		ReadContext:   resourceGitlabProjectProtectedEnvironmentRead,
+		DeleteContext: resourceGitlabProjectProtectedEnvironmentDelete,
+		Importer: &schema.ResourceImporter{
+			StateContext: schema.ImportStatePassthroughContext,
+		},
+		Schema: map[string]*schema.Schema{
+			"project": {
+				Description:  "The ID or full path of the project which the protected environment is created against.",
+				Type:         schema.TypeString,
+				ForceNew:     true,
+				Required:     true,
+				ValidateFunc: validation.StringIsNotEmpty,
+			},
+			"environment": {
+				Description:  "The name of the environment.",
+				Type:         schema.TypeString,
+				ForceNew:     true,
+				Required:     true,
+				ValidateFunc: validation.StringIsNotEmpty,
+			},
+			"deploy_access_levels": {
+				Description: "Array of access levels allowed to deploy, with each described by a hash.",
+				Type:        schema.TypeList,
+				ForceNew:    true,
+				Required:    true,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"access_level": {
+							Description:  "Level of access required to deploy to this protected environment (developer, maintainer)",
+							Type:         schema.TypeString,
+							ForceNew:     true,
+							Optional:     true,
+							ValidateFunc: validation.StringInSlice([]string{"developer", "maintainer"}, false),
+						},
+						"access_level_description": {
+							Description: "Readable description of level of access.",
+							Type:        schema.TypeString,
+							Computed:    true,
+						},
+						"user_id": {
+							Description:  "The ID of the user allowed to deploy to this protected environment.",
+							Type:         schema.TypeInt,
+							ForceNew:     true,
+							Optional:     true,
+							ValidateFunc: validation.IntAtLeast(1),
+						},
+						"group_id": {
+							Description:  "The ID of the group allowed to deploy to this protected environment.",
+							Type:         schema.TypeInt,
+							ForceNew:     true,
+							Optional:     true,
+							ValidateFunc: validation.IntAtLeast(1),
+						},
+					},
+				},
+			},
+		},
+	}
+})
+
+func resourceGitlabProjectProtectedEnvironmentCreate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
```

Comment:
For these functions there are a lot of the same feedbacks as I already gave in the `gitlab_project_environment` resource, like with the context, usage of `gitlab.String()`, `is404(err)`, ...

Can you address them here as well to your best knowledge and then I'll review that part again?

Reply:
I believe I have addressed most/all of these occurrences now in the protected environment resource